src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -0,0 +1,308 @@
+    uint32 i;
+    for(i=0;i<req_sz;i++){
+        buf[i]=start[i+offset];
+    }
+    return OS_FS_SUCCESS;
+}
+
+/*
+ * Copies information from a tar file header to a fsobj
+ */
+uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file)
+{
+    //TODO: store filenames without paths
+    //TODO: more descriptive error checking
+    assert(tar_offset<size);
+    assert(file->filedes>0);
+
+    char *location = start;
+    location += tar_offset;
+
+    file -> name        = location;
+    if(*(location+strlen(location)-1)=='/'){
+        file -> type    = FSOBJ_DIR;
+        file -> size    = 0;
+    }
+    else{ 
+        file -> type    = FSOBJ_FILE;
+        file -> size    = oct_to_dec(location+124);
+        file -> data    = location+500;
+    }
+    return OS_FS_SUCCESS;
+}
+
+//this could be more performate bithack
+uint32 round_up(uint32 offset)     
+{
+    if(offset%512) return offset + (512-(offset%512));
+    return offset;
+}
+
+/******************************************************************************
+** fsobj Level Methods
+******************************************************************************/
+
+uint32 fsobj_init(struct fsobj *o)
+{
+    assert(o);
+    assert(o->filedes>0);
+	
+    o->name     = "EXAMPLE_NAME";
+	o->type     = FSOBJ_DIR;
+	o->size     = 0;
+    o->refcnt   = 0;
+	o->data     = NULL;
+	o->child    = o->parent = NULL;
+    o->next     = o->prev = NULL;
+    return OS_FS_SUCCESS;
+}
+
+//finds the next open file
+uint32 newfile_get(struct fsobj **o)
+{
+    uint32 count=0;
+    while(count<MAX_NUM_FILES && files[count].filedes){
+        count++;
+    }
+    if(count==MAX_NUM_FILES) return OS_FS_ERROR;
+    *o =  &files[count];
+    (*o) -> filedes=count+1; //filedes needs to be unique and nonzero, so use index
+    return OS_SUCCESS;
+}
+
+uint32 file_insert(struct fsobj *o, struct FS *filesys){
+    assert(filesys && filesys->root && filesys->root->filedes);
+    if(filesys->root==o) return OS_SUCCESS;
+    
+    insert(o,filesys->root);
+
+    return OS_SUCCESS;
+}
+
+//could be written iteratively to use less stackframe if that becomes an issue
+// Root is 
+uint32 insert(struct fsobj *o, struct fsobj *root){
+
+    assert(root);
+    assert(strlen(root->name)<strlen(o->name));
+    assert(strcmp(root->name,o->name)!=0);
+    assert(strspn(o->name,root->name) >= strlen(root->name));
+
+    if(!root->child){
+        o->parent=root;
+        root->child=o;
+        return OS_SUCCESS;
+    }
+    root=root->child;
+    
+    do{
+        if(strspn(o->name,root->name) >= strlen(root->name)) return insert(o, root);
+        if(root->next) root = root->next;
+    } while(root->next);
+
+    root->next = o;
+    o->prev = root;
+    o->parent = root->parent;
+    return OS_SUCCESS;
+}
+
+int32 open(const char *path){
+    assert( fs[0].root && !fs[1].root );
+    struct fsobj *o=fs[0].root;
+    o=find(path,o);
+    if(o){
+        o->refcnt++;
+        return o->filedes;
+    }
+    return OS_FS_ERROR;
```

Comment:
I'd reorder this so you bail out on `!o`. I just think it makes your intent more clear.